pandas/_libs/index.pyx

```diff
@@ -712,14 +712,16 @@ cdef class BaseMultiIndexCodesEngine:
             Pre-calculated offsets, one for each level of the index.
         """
         self.levels = levels
-        self.offsets = offsets
+        # Downcast the type if possible, to prevent upcasting when shifting codes:
+        self.offsets = offsets.astype(np.min_scalar_type(offsets[0]), copy=False)
```

Comment:
Can this be done in `MultiIndex._engine` instead? There's already an `astype("uint64")` there

Reply:
Sure, makes sense